gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandler.java

```diff
@@ -66,16 +292,55 @@ private String parseFromAsid(COPCIN000001UK01Message payload) {
             .get(0)
             .getExtension();
     }
-
     private String parseToAsid(COPCIN000001UK01Message payload) {
         return payload.getCommunicationFunctionSnd()
             .getDevice()
             .getId()
             .get(0)
             .getExtension();
     }
-
     private String parseMessageRef(COPCIN000001UK01Message payload) {
         return payload.getId().getRoot();
     }
+    public boolean sendNackMessage(NACKReason reason, COPCIN000001UK01Message payload, String conversationId) {
+
+        LOGGER.debug("Sending NACK message with acknowledgement code [{}] for message EHR Extract message [{}]", reason.getCode(),
+            payload.getId().getRoot());
+
+        migrationStatusLogService.addMigrationStatusLog(reason.getMigrationStatus(), conversationId);
+
+        return sendNACKMessageHandler.prepareAndSendMessage(prepareNackMessageData(
+            reason,
+            payload,
+            conversationId
+        ));
+    }
+    private NACKMessageData prepareNackMessageData(NACKReason reason, COPCIN000001UK01Message payload,
+        String conversationId) {
+
+        String toOdsCode = parseToOdsCode(payload);
+        String messageRef = parseMessageRef(payload);
+        String toAsid = parseToAsid(payload);
+        String fromAsid = parseFromAsid(payload);
+        String nackCode = reason.getCode();
+
+        return NACKMessageData.builder()
+            .conversationId(conversationId)
+            .nackCode(nackCode)
+            .toOdsCode(toOdsCode)
+            .messageRef(messageRef)
+            .toAsid(toAsid)
+            .fromAsid(fromAsid)
+            .build();
+    }
+    private String parseToOdsCode(COPCIN000001UK01Message payload) {
+
+        Element gp2gpElement = payload.getControlActEvent()
+            .getSubject()
+            .getPayloadInformation()
+            .getValue()
+            .getAny()
+            .get(0);
+        return gp2gpElement.getFirstChild().getNextSibling().getNextSibling().getFirstChild().getNodeValue();
```

Comment:
i don't love this... what is going on? can we make this more readable and/or robust to changes to the ordering? if not, a careful explanatory comment is required.